tuf/formats.py

```diff
@@ -232,8 +232,8 @@
 # and sha512 may be computed for the same file and stored).
 FILEINFO_SCHEMA = SCHEMA.Object(
   object_name = 'FILEINFO_SCHEMA',
-  length = LENGTH_SCHEMA,
-  hashes = HASHDICT_SCHEMA,
+  length = SCHEMA.Optional(LENGTH_SCHEMA),
```

Comment:
Not sure this is a good idea. Because in `targets`, they are necessary, but not so in `timestamp` or `snapshot`, so need to distinguish between these two schemas.

Reply:
Absolutely, as @trishankatdatadog has said. From the specification we can see that:
* in _snapshot.json_, lengths and hashes of the targets metadata files (top-level and delegations) are optional
* in _targets.json_ (and delegated targets metadata), lengths and hashes of the target files are required
* in _timestamp.json_, lengths and hashes of the snapshot.json file are optional

Perhaps we can add a `METADATA_FILEINFO_SCHEMA` to formats.py for use here?

Reply:
Yes, if so, perhaps rename the original to `TARGETS_FILEINFO_SCHEMA`, just to clarify

Reply:
I will rename the original `FILE_INFO_SCHEMA` to `TARGETS_FILEINFO_SCHEMA` as suggested, but there is no need to add an additional `METADATA_FILEINFO_SCHEMA`.
There is one schema created by you @joshuagl around 2 months ago: 
```
LOOSE_FILEINFO_SCHEMA = SCHEMA.Object(
  object_name = "LOOSE_FILEINFO_SCHEMA",
  length = SCHEMA.Optional(LENGTH_SCHEMA),
  hashes = SCHEMA.Optional(HASHDICT_SCHEMA),
  version = SCHEMA.Optional(METADATAVERSION_SCHEMA),
  custom = SCHEMA.Optional(SCHEMA.Object())
)
```
https://github.com/theupdateframework/tuf/blob/d7aec6a5f9bbf0eafb7f075e3286d9e8db4c92c9/tuf/formats.py#L307 which could do the work.

Reply:
Version is not optional for snapshot and timestamp, therefore `LOOSE_FILEINFO_SCHEMA` could allow incorrectly formatted metadata.

Aside: I'm not a fan of `LOOSE_FILEINFO_SCHEMA`, I'm hoping we can remove it as we work on refactoring tuf.

Reply:
Yes, you are right. Version is not optional for timestamp and snapshot, it seems I missed it because those two look similar.
It seems I made the same mistake with the initial changes too.

Then, I will create a new schema. I was thinking of a name like `FILEINFO_OPTIONAL_FIELDS_SCHEMA` but it's too long...
Other naming suggestions?

Reply:
I'd propose we:
* rename FILEINFO_SCHEMA -> TARGETS_FILEINFO_SCHEMA
* rename LOOSE_FILEINFO_SCHEMA -> LOOSE_TARGETS_FILEINFO_SCHEMA
* add a new METADATA_FILEINFO_SCHEMA with optional length, optional hashes and version (not optional) fields.

Reply:
Yes, I agree with those three points and I will add them.

Reply:
@joshuagl I addressed your comments in a new commit.